# DP: Fix PR tree-optimization/54563

gcc/

2012-09-20  Jakub Jelinek  <jakub@redhat.com>

	Backported from mainline
	2012-09-17  Jakub Jelinek  <jakub@redhat.com>

	PR tree-optimization/54563
	* tree-ssa-math-opts.c (execute_cse_sincos): Call
	gimple_purge_dead_eh_edges if last call has been changed.

gcc/testsuite/

2012-09-20  Jakub Jelinek  <jakub@redhat.com>

	Backported from mainline
	2012-09-17  Jakub Jelinek  <jakub@redhat.com>

	PR tree-optimization/54563
	* g++.dg/torture/pr54563.C: New test.

--- a/src/gcc/testsuite/g++.dg/torture/pr54563.C	(revision 0)
+++ b/src/gcc/testsuite/g++.dg/torture/pr54563.C	(revision 191571)
@@ -0,0 +1,14 @@
+// PR tree-optimization/54563
+// { dg-do compile }
+
+extern "C" float powf (float, float);
+struct S { ~S (); };
+double bar ();
+double x;
+
+void
+foo ()
+{
+  S s;
+  x = powf (bar (), 2.);
+}
--- a/src/gcc/tree-ssa-math-opts.c	(revision 191570)
+++ b/src/gcc/tree-ssa-math-opts.c	(revision 191571)
@@ -1,5 +1,5 @@
 /* Global, SSA-based optimizations using mathematical identities.
-   Copyright (C) 2005, 2006, 2007, 2008, 2009, 2010, 2011
+   Copyright (C) 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012
    Free Software Foundation, Inc.
 
 This file is part of GCC.
@@ -1387,12 +1387,18 @@
   FOR_EACH_BB (bb)
     {
       gimple_stmt_iterator gsi;
+      bool cleanup_eh = false;
 
       for (gsi = gsi_after_labels (bb); !gsi_end_p (gsi); gsi_next (&gsi))
         {
 	  gimple stmt = gsi_stmt (gsi);
 	  tree fndecl;
 
+	  /* Only the last stmt in a bb could throw, no need to call
+	     gimple_purge_dead_eh_edges if we change something in the middle
+	     of a basic block.  */
+	  cleanup_eh = false;
+
 	  if (is_gimple_call (stmt)
 	      && gimple_call_lhs (stmt)
 	      && (fndecl = gimple_call_fndecl (stmt))
@@ -1430,6 +1436,7 @@
 		      gimple_set_location (new_stmt, loc);
 		      unlink_stmt_vdef (stmt);
 		      gsi_replace (&gsi, new_stmt, true);
+		      cleanup_eh = true;
 		    }
 		  break;
 
@@ -1450,6 +1457,7 @@
 		      gimple_set_location (new_stmt, loc);
 		      unlink_stmt_vdef (stmt);
 		      gsi_replace (&gsi, new_stmt, true);
+		      cleanup_eh = true;
 		    }
 		  break;
 
@@ -1465,6 +1473,7 @@
 		      gimple_set_location (new_stmt, loc);
 		      unlink_stmt_vdef (stmt);
 		      gsi_replace (&gsi, new_stmt, true);
+		      cleanup_eh = true;
 		    }
 		  break;
 
@@ -1472,6 +1481,8 @@
 		}
 	    }
 	}
+      if (cleanup_eh)
+	cfg_changed |= gimple_purge_dead_eh_edges (bb);
     }
 
   statistics_counter_event (cfun, "sincos statements inserted",
